pkg/cli/cert/cert.go

```diff
@@ -0,0 +1,237 @@
+package cert
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/erikdubbelboer/gspt"
+	"github.com/otiai10/copy"
+	"github.com/rancher/k3s/pkg/cli/cmds"
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/datadir"
+	"github.com/rancher/k3s/pkg/server"
+	"github.com/rancher/k3s/pkg/version"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+const (
+	adminComponent             = "admin"
+	apiServerComponent         = "api-server"
+	controllerManagerComponent = "controller-manager"
+	schedulerComponent         = "scheduler"
+	etcdComponent              = "etcd"
+	programControllerComponent = "-controller"
+	authProxyComponent         = "auth-proxy"
+	cloudControllerComponent   = "cloud-controller"
+	kubeletComponent           = "kubelet"
+	kubeProxyComponent         = "kube-proxy"
+)
+
+func commandSetup(app *cli.Context, cfg *cmds.Server, sc *server.Config) (string, string, error) {
+	gspt.SetProcTitle(os.Args[0])
+
+	nodeName := app.String("node-name")
+	if nodeName == "" {
+		h, err := os.Hostname()
+		if err != nil {
+			return "", "", err
+		}
+		nodeName = h
+	}
+
+	os.Setenv("NODE_NAME", nodeName)
+
+	sc.ControlConfig.DataDir = cfg.DataDir
+	sc.ControlConfig.Runtime = &config.ControlRuntime{}
+	dataDir, err := datadir.Resolve(cfg.DataDir)
+	if err != nil {
+		return "", "", err
+	}
+	return filepath.Join(dataDir, "server"), filepath.Join(dataDir, "agent"), err
+}
+
+func Run(app *cli.Context) error {
+	if err := cmds.InitLogging(); err != nil {
+		return err
+	}
+	return rotate(app, &cmds.ServerConfig)
+}
+
+func rotate(app *cli.Context, cfg *cmds.Server) error {
+	var serverConfig server.Config
+
+	serverDataDir, agentDataDir, err := commandSetup(app, cfg, &serverConfig)
+	if err != nil {
+		return err
+	}
+
+	serverConfig.ControlConfig.DataDir = serverDataDir
+	serverConfig.ControlConfig.Runtime = &config.ControlRuntime{}
+	deps.CreateRuntimeCertFiles(&serverConfig.ControlConfig, serverConfig.ControlConfig.Runtime)
+
+	tlsBackupDir, err := backupCertificates(serverDataDir, agentDataDir)
```

Comment:
It looks like we backup all the certificates, and then just remove the ones that are associated with the components we're rotating. Is there any reason we don't just backup the ones we're rotating?

What can be done with the backed-up certs and keys? Do we have another command to restore the certificates from backup if the rotation causes problems?

Reply:
well since this operation can be dangerous I opted for backing up everything everytime the user tries to run this command, plus adding a timestamp for more information, that said I didnt really add a behavior for restoring things, the backup was intended to be a fallback or disaster recovery kind of situation